### PR TITLE
feat: implement MCTS

### LINKED ISSUE #46: MCTS
https://github.com/michaelbzms/MonteCarloTreeSearch

### PR DESCRIPTION
**Pull request by Samarium150**
Close #46

### CODE REVIEW

**Review by pwyq — COMMENTED** [review]
which part is the rollout of MCTS

**Comment by Samarium150** [comment]
> which part is the rollout of MCTS

`simulate`

File: src/solver/mcts.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  while (!game.isTerminal()) {
+    auto moves = game.getPossibleMoves();
+    std::uniform_int_distribution<> dist(0, moves.size() - 1);
+    auto move = moves[dist(rng_)];
+    game.unsafePlay(move.first, move.second);
+  }
+  return root_->game_.toPlay() != game.toPlay();
+}
+
+solver::helper::Move MCTS::search() {
```
Comment (pwyq) [inline src/solver/mcts.cpp:65]: to make it consistent with other agents, we better change this to `void search()`, and have a member `Move best_move_`
`inline Move best_move() { return best_move_; }`

**Review by pwyq — CHANGES_REQUESTED** [review]

**Comment by pwyq** [comment]
```
You may not use a using-directive to make all names from a namespace available.

// Forbidden -- This pollutes the namespace.
using namespace foo;
```
https://google.github.io/styleguide/cppguide.html#Namespaces

I am changing every `using namespace solver::helper;`

**Comment by Samarium150** [comment]
> ```
> You may not use a using-directive to make all names from a namespace available.
> 
> // Forbidden -- This pollutes the namespace.
> using namespace foo;
> ```
> https://google.github.io/styleguide/cppguide.html#Namespaces
> 
> I am changing every `using namespace solver::helper;`

Did I use it somewhere? I totally forgot.

**Comment by pwyq** [comment]
How did you run MCTS

I am doing
```
      MCTS* mcts = new MCTS(game);
      res_move   = mcts->search();
```

I got
```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Cannot play at 2 2 with value 2, which is already filled. Full game = 22.*131*..2
```

**Comment by Samarium150** [comment]
> ```
> terminate called after throwing an instance of 'std::invalid_argument'
>   what():  Cannot play at 2 2 with value 2, which is already filled. Full game = 22.*131*..2
> ```

Probably something wrong with `simulate`

**Comment by pwyq** [comment]
PR #71  is ready. It will marge into this `46-mcts` branch

**Review by pwyq — APPROVED** [review]

**Pull request merged**